File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionMap.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;
import co.nlighten.jsontransform.JsonElementStreamer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Stream;

public class TransformerFunctionMap extends TransformerFunction {
    static final Logger logger = LoggerFactory.getLogger(TransformerFunctionMap.class);

    public TransformerFunctionMap() {
        super(FunctionDescription.of(
            Map.of(
            "to", ArgumentType.of(ArgType.Transformer).position(0).defaultIsNull(true)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        Stream<?> inputStream;
        Object to;
        if (context.has("to")) {
            var streamer = context.getJsonElementStreamer(null);
            if (streamer == null)
                return null;
            inputStream = streamer.stream();
            to = context.getJsonElement("to", false); // we don't transform definitions to prevent premature evaluation
        } else {
            // [ input, to ]
            var arr = context.getJsonArray(null, false); // we don't transform definitions to prevent premature evaluation
            if (arr == null)
                return null;
            var adapter = context.getAdapter();
            var inputEl = context.transform(context.getPathFor(0), adapter.get(arr, 0));
            if (!adapter.isJsonArray(inputEl)) {
                logger.warn("{} was not specified with an array of items", context.getAlias());
                return null;
            }
            inputStream = adapter.stream(inputEl);
            to = adapter.get(arr, 1);
        }
        var i = new AtomicInteger(0);
        return JsonElementStreamer.fromTransformedStream(context, inputStream
            .map(x -> context.transformItem(to, x, i.getAndIncrement()))
        );
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionMap.ts
import { AsyncSequence, asAsyncSequence } from "@wortise/sequency";
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import JsonElementStreamer from "../JsonElementStreamer";

class TransformerFunctionMap extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        to: { type: ArgType.Transformer, position: 0, defaultIsNull: true },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    let inputStream: AsyncSequence<any>;
    let to: any;
    if (context.has("to")) {
      const streamer = await context.getJsonElementStreamer(null);
      if (streamer == null) return null;
      inputStream = streamer.stream();
      to = await context.getJsonElement("to", false); // we don't transform definitions to prevent premature evaluation
    } else {
      // [ input, to ]
      const arr = await context.getJsonArray(null, false); // we don't transform definitions to prevent premature evaluation
      if (arr == null) return null;
      const inputEl = await context.transform(arr[0]);
      if (!Array.isArray(inputEl)) {
        console.warn(`${context.getAlias()} was not specified with an array of items`);
        return null;
      }
      inputStream = asAsyncSequence(inputEl);
      to = arr[1];
    }
    return JsonElementStreamer.fromTransformedStream(
      context,
      inputStream.mapIndexed((i, x) => context.transformItem(to, x, i)),
    );
  }
}

export default TransformerFunctionMap;
